### PR TITLE
Add `no-object-as-default-parameter` rule

### LINKED ISSUE #208: Rule proposal: No default parameter options
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/208)

I have seen this pattern of using default parameters with an object as default options values:

```js
const foo = (options = {a: false}) => {
	// …
};
```

This looks ok, but it's easy to miss that it only works with one option. If the user passes in an object, the whole default object will be replaced.

So here, the `b` option default value will be lost:

```js
const foo = (options = {a: false, b: true}) => {
	// …
};

foo({a: true});
```

I have seen this mistake a lot in the wild, so would be good to have a rule for it.

Example: https://github.com/lukechilds/byte-range/blob/3d154a9af0f827dfbb44c5acf32029cf1a549ccb/src/index.js#L3

The correct way to handle this is to use object spreading:

```js
const foo = options => {
	options = {
		a: false,
		b: true,
		...options
	};
};

foo({a: true});
```

The rule should fail on parameters named `options` or `opts` that have a default parameters value that is an object literal. The exception is an empty object, which should be allowed.

## Fail

```js
const foo = (options = {a: false}) => {
	// …
};
```

```js
const foo = (opts = {a: false, b: true}) => {
	// …
};
```

## Pass

```js
const foo = options => {
	options = {
		a: false,
		b: true,
		...options
	};
};
```

```js
const foo = (options = {}) => {
	// …
};
```

Would be nice to auto-fix this, but I don't think it's very easy, so it's not a requirement to resolve this issue.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $82.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($80.00)
- [0maxxam0 0maxxam0](https://issuehunt.io/u/0maxxam0) ($2.00)

### Submitted pull Requests
- [#633 Add `disallow-objects-as-default` rule](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/pull/633)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by medusalix**
Simply reports any default-assignments of objects to function parameters as described in [this](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/208#issuecomment-576647974) comment.

Fixes #208

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#208: Rule proposal: No default parameter options](https://issuehunt.io/repos/55832243/issues/208)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by fisker** [comment]
How about `no-object-as-default`

**Comment by fisker** [comment]
On second though, will people think this rule forbid this following syntax?

```js
export {
    object as default,
// ^
    foo as bar
};
```

**Comment by medusalix** [comment]
No, I don't think users will mix that up. The description is pretty clear about the rule's purpose.

**Comment by sindresorhus** [comment]
I would name the rule `no-object-as-default-parameter` to make it clearer.

File: docs/rules/no-object-as-default.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Disallow the use of objects as default parameters
+
+Default parameters should not be passed to a function through an object. The `foo = {a: false}` parameter works fine if only used with one option. As soon as additional options are added, you risk replacing the whole `foo = {a: false, b: true}` object when passing only one option: `{a: true}`. For this reason, object destructuring should be used instead.
```
Comment (sindresorhus) [inline docs/rules/no-object-as-default.md:3]:
```suggestion
Default parameters should not be passed to a function through an object literal. The `foo = {a: false}` parameter works fine if only used with one option. As soon as additional options are added, you risk replacing the whole `foo = {a: false, b: true}` object when passing only one option: `{a: true}`. For this reason, object destructuring should be used instead.
```

File: rules/no-object-as-default.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+module.exports = {
+	create,
+	meta: {
+		type: 'problem',
+		docs: {
+			url: getDocumentationUrl(__filename)
+		},
+		messages: {
+			[MESSAGE_ID]: 'Do not assign default object to parameter `{{parameter}}`.'
```
Comment (sindresorhus) [inline rules/no-object-as-default.js:33]:
```suggestion
			[MESSAGE_ID]: 'Do not use an object literal as default for parameter `{{parameter}}`.'
```

?

File: rules/no-object-as-default.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+const getDocumentationUrl = require('./utils/get-documentation-url');
+
+const MESSAGE_ID = 'noObjectAsDefault';
+
+const objectParameterSelector = [
+	'AssignmentPattern',
+	'[left.type="Identifier"]',
+	'[right.type="ObjectExpression"]',
+	'[right.properties.length>0]'
+].join('');
```
Comment (fisker) [inline rules/no-object-as-default.js:11]: If I remember correctly, the `AssignmentPattern` not only can use as parameter, maybe assignment default value? Anyway, we can use a stricter selector.

Comment (fisker) [reply rules/no-object-as-default.js:11]: Confirmed `const {foo={bar:1}} = baz;`

**Comment by medusalix** [comment]
> Please add tests about class/object methods.

Could you please elaborate?

File: rules/no-object-as-default-parameter.js [hunk 4]
```diff
@@ -4,6 +4,8 @@ const getDocumentationUrl = require('./utils/get-documentation-url');
 const MESSAGE_ID = 'noObjectAsDefaultParameter';
 
 const objectParameterSelector = [
+	':matches(FunctionDeclaration, ArrowFunctionExpression)',
```
Comment (fisker) [inline rules/no-object-as-default-parameter.js:7]: I think `:function` should work.

File: rules/no-object-as-default-parameter.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
+'use strict';
+const getDocumentationUrl = require('./utils/get-documentation-url');
+
+const MESSAGE_ID = 'noObjectAsDefaultParameter';
+
+const objectParameterSelector = [
+	':matches(FunctionDeclaration, ArrowFunctionExpression)',
+	'>',
+	'AssignmentPattern',
```
Comment (fisker) [inline rules/no-object-as-default-parameter.js:9]: Though another child(function.id) can't be `AssignmentPattern`, I still prefer use `AssignmentPattern.params`.

**Comment by fisker** [comment]
Please test


```js
class A {
  method(foo = {bar:1}) {}
}
```

```js
const A = class {
  method(foo = {bar:1}) {}
}
```

```js
object = {
  method(foo = {bar:1}) {}
}
```

File: rules/no-object-as-default-parameter.js (hunk 4 above)
Comment (fisker) [reply rules/no-object-as-default-parameter.js:7]: If all your tests passed, you must missed the `FunctionExpression`, pleaseadd one.

**Comment by medusalix** [comment]
> const foo = ({bar: baz = {a = false, b = 123}}) => {};

Is this valid syntax? It outputs `Shorthand property assignments are valid only in destructuring patterns`.

File: rules/no-object-as-default-parameter.js (hunk 4 above)
Comment (medusalix) [reply rules/no-object-as-default-parameter.js]: Yep, I've missed that one. Fixed.

**Comment by fisker** [comment]
`const adc = () => (foo = {bar: 1});`  This one looks similar to default parameter, but it's not , let's add it to test too.

**Comment by fisker** [comment]
> > const foo = ({bar: baz = {a = false, b = 123}}) => {};
> > const foo = ({foo = {a = false, b = 123}}) => {};
> 
> Is this syntactically correct? It gives me `Shorthand property assignments are valid only in destructuring patterns`.

Okay , if  it's not  valid, I  mean  to  test  deep destructuring.


> > Maybe also constructor and getter setter
> 
> I've added tests for the constructor and setter, it doesn't seem like getters allow parameters.

Okay.

**Comment by fisker** [comment]
Looks  good , I'll check  again when ai  back to my  laptop.

**Comment by medusalix** [comment]
Thank you. I really appreciate your constructive feedback, it's always very helpful 👍.

File: rules/no-object-as-default-parameter.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	'[left.type="Identifier"]',
+	'[right.type="ObjectExpression"]',
+	'[right.properties.length>0]'
+].join('');
+
+const create = context => {
+	return {
+		[objectParameterSelector]: node => {
+			context.report({
+				node,
```
Comment (fisker) [inline rules/no-object-as-default-parameter.js:17]: @medusalix  Do you think we should report on parameter name? `node.left`

Comment (medusalix) [reply rules/no-object-as-default-parameter.js]: Yes, makes sense.

**Comment by sindresorhus** [comment]
Great! Looks good to me too 👍

**Pull request merged**